File: LeetCode/TwoPointers/java/validPalindrome.java
package LeetCode.TwoPointers.java;

record validPalindromeRecord(String s, boolean expected) {
}

public class validPalindrome {
    public boolean isPalindrome(String s) {
        int l = 0;
        int r = s.length() - 1;

        while (l < r) {
            // move the l pointer until it points to a letter or digit
            while (l < r && !Character.isLetterOrDigit(s.charAt(l))) {
                l += 1;
            }

            // move the r pointer until it points to a letter or digit
            while (r > l && !Character.isLetterOrDigit(s.charAt(r))) {
                r -= 1;
            }

            if (Character.toLowerCase(s.charAt(l)) != Character.toLowerCase(s.charAt(r))) {
                return false;
            }

            l += 1;
            r -= 1;
        }

        return true;
    }

    public static void main(String[] args) {
        validPalindromeRecord input = new validPalindromeRecord("A man, a plan, a canal: Panama", true);
        validPalindrome.testSolution(input);

        input = new validPalindromeRecord("race a car", false);
        validPalindrome.testSolution(input);

        input = new validPalindromeRecord(" ", true);
        validPalindrome.testSolution(input);
    }

    private static void testSolution(validPalindromeRecord input) {
        System.out.println("Input: s: " + input.s());
        System.out.println("Expected: " + input.expected());
        boolean val = new validPalindrome().isPalindrome(input.s());
        System.out.println("Result: " + val + ", " + (val == input.expected() ? "Correct" : "Wrong"));
        System.out.println("-".repeat(50));
    }
}

File: LeetCode/TwoPointers/python/validPalindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        l = 0
        r = len(s) - 1

        while l < r:
            while l < r and not s[l].isalnum():
                l += 1

            while r > l and not s[r].isalnum():
                r -= 1

            if s[l].lower() != s[r].lower():
                return False
            
            l += 1
            r -= 1

        return True
    
    @staticmethod
    def testSolution(input) -> None:
        print("Input: s: {}".format(input["s"]))
        print("Expected: {}".format(input["expected"]))
        val = Solution().isPalindrome(input["s"])
        print("Result: {} , {}".format(val, ("Correct" if val == input["expected"] else "Wrong")))
        print("-" * 50)

if __name__ == "__main__":
    input = {"s":"A man, a plan, a canal: Panama", "expected":True}
    Solution.testSolution(input)

    input = {"s":"race a car", "expected":False}
    Solution.testSolution(input)

    input = {"s":" ", "expected":True}
    Solution.testSolution(input)

